### PR TITLE
native support for postgres array types (close #348)

### LINKED ISSUE #348: better support for postgres array/hstore types
Currently the input for an array is a literal, i.e for a column of type `text[]` the input value has to be a string as follows: `"{hello,world}"`. We should do the appropriate mapping to json arrays.

### PR DESCRIPTION
**Pull request by rakeshkky**
### Description

TODO:- [console] support new input types format (json) for functions. 

Allows following types of queries 
```
query q {
  bar: foo (where: { a : { _contained_in : [1,2,3,4,5,6,7,8]  }} ) {
    a
  }

  baz: foo (where: { a : { _contains : [1,2,3,4] }} ) {
    a
  } 
}
```
And the following mutations with array
```
mutation m {
  insert_foo ( objects: [ {a: [1,3,5,7,9] } ]){
    affected_rows
    returning {
      a 
      id
    }
  }
}
```

Support `_preprend` and `_append` operators in updating array columns

### Affected components 

- Server
- Tests
- Console

### Related Issues

#348 
#349

### Solution and Design

- `PGColType` is modified to support postgres types
- `PGColValue` is modified to parse json values according to new `PGColType`
-  `hdb_type_info` view is added to `hdb_catalog` schema to list all types with details
- catalog version bump

### CODE REVIEW

File: server/src-rsr/migrate_from_17_to_18.sql [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+  rtn.nspname::text AS return_type_schema,
+  rt.typname::text AS return_type_name,
+
+  CASE
+    WHEN ((rt.typtype) :: text = ('b' :: character(1)) :: text) THEN 'BASE' :: text
+    WHEN ((rt.typtype) :: text = ('c' :: character(1)) :: text) THEN 'COMPOSITE' :: text
+    WHEN ((rt.typtype) :: text = ('d' :: character(1)) :: text) THEN 'DOMAIN' :: text
+    WHEN ((rt.typtype) :: text = ('e' :: character(1)) :: text) THEN 'ENUM' :: text
+    WHEN ((rt.typtype) :: text = ('r' :: character(1)) :: text) THEN 'RANGE' :: text
+    WHEN ((rt.typtype) :: text = ('p' :: character(1)) :: text) THEN 'PSUEDO' :: text
```
Comment (lorenzo) [inline server/src-rsr/migrate_from_17_to_18.sql]: Is `PSUEDO` a typo? Should it be `PSEUDO`?

**Comment by dsandip** [comment]
@jberryman some more context on this https://hasurahq.slack.com/archives/CKFUG6RCH/p1562793618269300

**Comment by cthurston** [comment]
Will this patch allow for array relationships to be defined from a uuid[] array field?

I tried to define a relationship from a uuid[] field to another table:
````
    "hint": "No operator matches the given name and argument type(s). You might need to add explicit type casts.",
     "message": "operator does not exist: uuid[] = uuid",
````

**Comment by andrewalex** [comment]
Also related (From the permissions side) #2882

**Comment by himat** [comment]
What is the status of this?

**Comment by karibertils** [comment]
Looking forward to this one

**Comment by maxcan** [comment]
Is it safe to assume that this current PR is dead?  @marionschleifer,  @0x777 any idea if native array support will be on the roadmap soon? 

My hasura instances are littered with array sub tables since I prefer the strict type safety that gives me over JSONB fields.

**Comment by s-kris** [comment]
Looking forward to this

**Comment by 0x777** [comment]
Hey folks, I'm sorry about letting this PR go stale (preserving backwards compatibility with the above changes is very hard). Over the past few weeks, we've been working on some changes to the codebase that'll help us improve the speed at which we can implement new features. As soon as we are done with it, we'll pick up these PRs.

**Comment by maxcan** [comment]
@0x777 appreciate the update!  thanks!

**Comment by xiaoyu-tamu** [comment]
What is the status of this?

**Comment by maxcan** [comment]
@coco98 ?

**Comment by axhl** [comment]
Any updates on where this sits on the roadmap? Lots of folks are excited about it: https://github.com/hasura/graphql-engine/issues/348

**Comment by Fi1osof** [comment]
2021 year. Still waiting for.

**Comment by karibertils** [comment]
2022 and still waiting

**Pull request closed without merging**